doc/manual/typing.md

```diff
@@ -0,0 +1,559 @@
+that determining the apparent type shouldn't recurse arbitrarily inside the
+expression or do anything non-trivial.  Typically, our previous `0 + 1` is a
+compound expression with a primitive operator, so the apparent type is just
+`Dyn`. For now, the typechecker determine an apparent type that is not `Dyn` for
+literals (numbers, strings, booleans), variables, and annotated expressions.
+
+Otherwise, the typechecker fallbacks to `Dyn`. In the future, it could also
+infer `Dyn -> Dyn` for functions, `{_: Dyn}` for records, and so on. As of now
+(INSERT VERSION HERE), it doesn't.
+
+We could add a type annotation to `x`. But sometimes we don't want to, or we
+can't. Maybe we don't want the typechecker to do its ceremony and recurse in the
+whole expression. Maybe the expression is not typeable in Nickel, but we still
+know it will evaluate to a number. In any of these situations, we can trade the
+type annotation for a contract application:
+
+Exmaple:
+```
+let x | Num = if true then 0 else "a" in
+(1 + x : Num)
+```
+
+Here, `x` is clearly always a number, but it is not well-typed (the `then` and
+`else` branches of an `if` must have the same type). Nonetheless, this program
+is accepted! The rationale is that because we inserted a contract application,
+the typechecker knows that if `x` is not a number, the program will fail early
+with a detailed contract error. Thus, if we reach `1 + x`, `x` is necessarily a
+number at this point, and everything will be fine. In a way, the contract
+application acts like a type cast, but whose verification is delayed to
+run-time.
+
+When using a contract annotation inside typed code, the *typechecker is turned
+off again*. Contract application is dual to type annotation with respect to
+typechecking. This is illustrated by the following variation being accepted:
+
+```nickel
+(1 + ((if true then 0 else "a" | Num)) : Num
+```
+
+While, as said before, this one is rejected:
+
+```nickel
+(1 + (if true then 0 else "a")) : Num
+```
+
+Result:
+```
+error: Incompatible types
+  ┌─ repl-input-46:1:27
+  │
+1 │ (1 + (if true then 0 else "a")) : Num
+  │                           ^^^ this expression
+  │
+  = The type of the expression was expected to be `Num`
+  = The type of the expression was inferred to be `Str`
+  = These types are not compatible
+```
+
+**Take-away**: When calling to typed code from untyped code, Nickel
+automatically inserts contract checks at the boundary to enjoy clearer and
+earlier error reporting. In the other direction, an expression `exp | Type` is
+blindly accepted to be of type `Type` by the typechecker. This is a way of using
+untyped values inside typed code by telling the typechecker "trust me on this
+one, and if I'm wrong there will be a contract error anyway". Finally, while a
+type annotation switches the typechecker on, a contract annotation switches it
+back off.
+
+If you read until this point, you may feel confused. We have static types on one
```

Comment:
I certainly hope that the reader isn't confused, otherwise the documentation hasn't done its job. At any rate, better not suggest that they do.

Reply:
Got rid of this paragraph.